nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/PublisherLease.java

```diff
@@ -193,10 +196,24 @@ void publish(final FlowFile flowFile, final RecordSet recordSet, final RecordSet
                     additionalAttributes = writeResult.getAttributes();
                     writer.flush();
                 }
-
                 final byte[] messageContent = baos.toByteArray();
-                final String key = messageKeyField == null ? null : record.getAsString(messageKeyField);
-                final byte[] messageKey = (key == null) ? null : key.getBytes(StandardCharsets.UTF_8);
+
+                final byte[] messageKey;
+                if ((recordKeyWriterFactory == null) || (messageKeyField == null)) {
+                    messageKey = Optional.ofNullable(record.getAsString(messageKeyField))
+                            .map(s -> s.getBytes(StandardCharsets.UTF_8)).orElse(null);
+                } else {
+                    try (final ByteArrayOutputStream os = new ByteArrayOutputStream(1024)) {
+                        final Record keyRecord = Optional.ofNullable(record.getValue(messageKeyField))
+                                .filter(Record.class::isInstance).map(Record.class::cast)
+                                .orElseThrow(() ->  new IOException("The property 'Record Key Writer' is defined, but the record key is not a record"));
```

Comment:
Not sure that it makes sense to throw an Exception in this case. If the key is a `String`, we should just convert it to a `byte[]` via `String.toByteArray(StandardCharsets.UTF_8)`. If the key is a `byte[]` already then just send it as-is. We should only use the Record Reader when necessary.
In short - we should be explicit in what we produce, and lenient in what we accept.

Reply:
I think I would actually do something like:
```
final Object key;
final Object keyValue = record.getValue(messageKeyField);
if (keyValue == null) {
  key = keyValue;
} else if (keyValue instanceof byte[]) {
  key = (byte[]) keyValue;
} else if (keyValue instanceof Byte[]) {
  // This case exists because in our Record API we currently don't have a BYTES type, we use an Array of type Byte, which creates a Byte[] instead of a byte[]. We should address this in the future, but we should account for the log here.
  final Byte[] bytes = (Byte[]) keyValue;
  key = new byte[bytes.length];
  for (int i=0; i < bytes.length; i++) {
    key[i] = bytes[i];
  }
} else if (key instanceof Record) {
 ... // what you have below, to create record writer, write key, flush
} else {
  final String keyString = keyValue.toString();
  key = keyString.getBytes(StandardCharsets.UTF_8);
}
```
I.e., just look at the object that you have. Using the schema is also feasible. But it can get really complicated if the schema indicates that it's a CHOICE type because then you'll probably still end up having to do this type of logic.